azpubsub/contextvalidator/src/main/java/org/apache/kafka/common/security/authenticator/AzPubSubPrincipalBuilder.java

```diff
@@ -0,0 +1,136 @@
+package org.apache.kafka.common.security.authenticator;
+
+import azpubsub.kafka.security.authenticator.AzPubSubPrincipal;
+import azpubsub.kafka.security.authenticator.SaslAuthenticationContextValidator;
+import azpubsub.kafka.security.authenticator.SslAuthenticationContextValidator;
+import kafka.server.KafkaConfig;
+import org.apache.kafka.common.errors.IllegalSaslStateException;
+import org.apache.kafka.common.errors.InvalidConfigurationException;
+import org.apache.kafka.common.security.auth.*;
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.naming.AuthenticationNotSupportedException;
+import javax.security.sasl.AuthenticationException;
+import javax.security.sasl.SaslServer;
+import java.util.Map;
+
+
+/**
+ *  A new Principal Builder used to replace the default provider.
+ *  This provider contructs both regular principal (Kafka Principal) and also SAML token principal.
+ */
+public class AzPubSubPrincipalBuilder implements KafkaPrincipalBuilder, Configurable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AzPubSubPrincipalBuilder.class);
+    private Class<?> sslAuthenticationContextValidatorClass = null;
+    private Class<?> saslAuthenticationContextValidatorClass = null;
+    private SslAuthenticationContextValidator sslAuthenticationContextValidator = null;
+    private SaslAuthenticationContextValidator saslAuthenticationContextValidator = null;
+
+    public void configure(Map<String, ?> configs) {
+        try{
+            sslAuthenticationContextValidatorClass = Class.forName(configs.get(KafkaConfig.AzpubsubSslAuthenticationValidatorClassProp()).toString());
```

Comment:
Class.forName breaks encapsulation. In general, we should not use reflection in Java unit tests. In other words, we should write code that is testable.

Reply:
We have to do this: class.ForName(). because the validator class is a plugin configured from server.properties.  And this class is loaded once for each connection, not for each request. Performance is slightly impacted.